heat/core/manipulations.py

```diff
@@ -735,6 +736,170 @@ def hstack(tup):
+
+    def reshape_argsort_counts_displs(shape1, lshape1, displs1, shape2, displs2, axis, comm):
+        """
+        Compute the send order, counts, and displacements.
+        """
+        mask = torch.zeros(lshape1).flatten()
+
+        local_index = 0
+        global_index = displs1[comm.rank] * np.prod(shape1[axis + 1 :])
+        local_len = np.prod(lshape1[axis:])
+        global_len = np.prod(shape1[axis:])
+
+        ulen = np.prod(shape2[axis + 1 :])
+
+        while local_index < np.prod(lshape1):
+            for i in range(local_len):
+                pos = global_index + i
+                jpos = (pos // ulen) % shape2[axis]
+                # search new process id
+                for j, k in zip(displs2, range(comm.size)):
+                    if j <= jpos:
+                        mask[local_index + i] = k
+            global_index += global_len
+            local_index += local_len
+
+        uniques, counts = torch.unique(mask, return_counts=True)
+
+        mpicounts = torch.zeros(comm.size)
+
+        for i, j in zip(uniques, counts):
+            mpicounts[int(i)] = int(j)
+
+        return (
+            mask.argsort(),
+            tuple(mpicounts.numpy()),
+            (0,) + tuple(torch.cumsum(mpicounts[:-1], dim=0).numpy()),
+        )
+
+    # Check the type of shape and number elements
+    shape = stride_tricks.sanitize_shape(shape)
+    if np.prod(shape) != a.size:
+        raise ValueError("cannot reshape array of size {} into shape {}".format(a.size, shape))
+
+    # Forward to Pytorch directly
+    if a.split is None:
+        return factories.array(
+            torch.reshape(a._DNDarray__array, shape), dtype=a.dtype, device=a.device, comm=a.comm
+        )
+
+    # Create new flat result tensor
+    _, local_shape, _ = a.comm.chunk(shape, a.split)
+    data = torch.empty(
+        local_shape, dtype=a.dtype.torch_type(), device=a.device.torch_device
+    ).flatten()
+
+    # Calculate the counts and displacements
+    _, old_displs, _ = a.comm.counts_displs_shape(a.shape, a.split)
+    _, new_displs, _ = a.comm.counts_displs_shape(shape, a.split)
+
+    if a.split > 0:
+        sendsort, sendcounts, senddispls = reshape_argsort_counts_displs(
+            a.shape, a.lshape, old_displs, shape, new_displs, a.split, a.comm
+        )
+        recvsort, recvcounts, recvdispls = reshape_argsort_counts_displs(
+            shape, local_shape, new_displs, a.shape, old_displs, a.split, a.comm
+        )
+
+        # rearange order
+        send = a._DNDarray__array.flatten()[sendsort]
+        a.comm.Alltoallv((send.flatten(), sendcounts, senddispls), (data, recvcounts, recvdispls))
+
+        # original order
+        backsort = torch.argsort(recvsort)
+        data = data[backsort]
+
+    else:
+        # turn displs into cumulative sums
+        old_csum = tuple(np.prod(a.shape[1:]) * i for i in old_displs[1:]) + (a.size,)
+        new_csum = tuple(np.prod(shape[1:]) * i for i in new_displs[1:]) + (a.size,)
```

Comment:
torch not numpy

Reply:
There are now a bunch of torch function calls